swarm/src/behaviour.rs

```diff
@@ -123,37 +116,133 @@ pub(crate) type THandlerOutEvent<THandler> =
-    /// and the behaviour can send a message to the handler by making [`NetworkBehaviour::poll`]
-    /// return [`NetworkBehaviourAction::NotifyHandler`].
+    /// This is invoked once another peer has successfully dialed us.
     ///
-    /// Note that the handler is returned to the [`NetworkBehaviour`] on connection failure and
-    /// connection closing.
-    fn new_handler(&mut self) -> Self::ConnectionHandler;
+    /// At this point, we have verified their [`PeerId`] and we know, which particular [`Multiaddr`] succeeded in the dial.
+    /// In order to actually use this connection, this function must return a [`ConnectionHandler`].
+    /// Returning an error will immediately close the connection.
+    fn handle_established_inbound_connection(
+        &mut self,
+        peer: PeerId,
+        _connection_id: ConnectionId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<
+        <Self::ConnectionHandler as IntoConnectionHandler>::Handler,
+        Box<dyn std::error::Error + Send + 'static>,
+    > {
+        #[allow(deprecated)]
+        Ok(self.new_handler().into_handler(
+            &peer,
+            &ConnectedPoint::Listener {
+                local_addr: local_addr.clone(),
+                send_back_addr: remote_addr.clone(),
+            },
+        ))
+    }
+
+    /// Callback that is invoked for every outbound connection attempt.
+    ///
+    /// We have access to:
+    ///
+    /// - The [`PeerId`], if known. Remember that we can dial without a [`PeerId`].
+    /// - All addresses passes to [`DialOpts`] are passed in here too.
+    /// - The effective [`Role`](Endpoint) of this peer in the dial attempt. Typically, this is set to [`Endpoint::Dialer`] except if we are attempting a hole-punch.
+    /// - The [`ConnectionId`] identifying the future connection resulting from this dial, if successful.
+    ///
+    /// Any error returned from this function will immediately abort the dial attempt.
+    fn handle_pending_outbound_connection(
+        &mut self,
+        maybe_peer: Option<PeerId>,
+        _addresses: &[Multiaddr],
+        _effective_role: Endpoint,
+        _connection_id: ConnectionId,
+    ) -> Result<Vec<Multiaddr>, Box<dyn std::error::Error + Send + 'static>> {
+        if let Some(peer_id) = maybe_peer {
+            Ok(self.addresses_of_peer(&peer_id))
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    /// Callback that is invoked for every established outbound connection.
+    ///
+    /// This is invoked once we have successfully dialed a peer.
+    /// At this point, we have verified their [`PeerId`] and we know, which particular [`Multiaddr`] succeeded in the dial.
+    /// In order to actually use this connection, this function must return a [`ConnectionHandler`].
+    /// Returning an error will immediately close the connection.
+    fn handle_established_outbound_connection(
```

Comment:
Checking my understanding, we still need the `FromSwarm::ConnectionEstablished` event as any `NetworkBehaviour` within the hierarchy of behaviours might deny the connection.

If so, what do you think of elaborating this in the methods doc comment? Would that confuse the user more than it helps?

Reply:
Yes, the event indicates that the connection was not denied and can actually be used.